lib/mix/lib/mix/tasks/test/stale.ex

```diff
@@ -0,0 +1,261 @@
+defmodule Mix.Tasks.Test.Stale do
+  @moduledoc false
+
+  require Mix.Compilers.Elixir
+  alias Mix.Compilers.Elixir, as: CE
+
+  import Record
+
+  defrecordp :source, [
+    source: nil,
+    compile_references: [],
+    runtime_references: [],
+    external: []
+  ]
+
+  @stale_manifest ".compile.test_stale"
+  @manifest_vsn :v1
+
+
+  @doc """
+  Set up the `--stale` run for given matched test files, test paths, and options.
+
+  If there are tests to run, a tuple containing the test files to run, the PID
+  of the agent used in the each_module callback, and the callback options for
+  ParallelRequire is returned.
+
+  If there are no tests to run, the pid and callbacks are not returned, since the
+  test task should not perform a ParallelRequire if there are no tests to run.
+
+  The test task is responsible for running the returned tests, using the returned
+  ParallelRequire callbacks, calling the agent_write_manifest/1 function with the
+  provided PID after performing the ParallelRequire sucessfully, and calling the
+  agent_stop/1 function with the same PID after the ParallelRequire, whether it
+  succeeds or not.
+  """
+  def set_up(matched_test_files, test_paths, opts) do
+    manifest = manifest()
+    modified = Mix.Utils.last_modified(manifest)
+    all_sources = read_manifest()
+
+    removed =
+      for source(source: source) <- all_sources,
+          not(source in matched_test_files),
+          do: source
+
+    configs = Mix.Project.config_files
+    force = opts[:force] || Mix.Utils.stale?(configs, [manifest]) || test_helper_stale?(test_paths)
+
+    changed =
+      if force do
+        # let's just require everything
+        matched_test_files
+      else
+        sources_mtimes = mtimes(all_sources)
+
+        # Otherwise let's start with the new sources
+        for(source <- matched_test_files,
+            not List.keymember?(all_sources, source, source(:source)),
+            do: source)
+          ++
+        # Plus the sources that have changed in disk
+        for(source(source: source, external: external) <- all_sources,
+            times = Enum.map([source | external], &Map.fetch!(sources_mtimes, &1)),
+            Mix.Utils.stale?(times, [modified]),
+            do: source)
+      end
+
+    stale   = MapSet.new(changed -- removed)
+    sources = update_stale_sources(all_sources, removed, changed)
+
+    test_files_to_run =
+      sources
+      |> tests_with_changed_references()
+      |> MapSet.union(stale)
+      |> MapSet.to_list()
+
+    if test_files_to_run == [] do
+      write_manifest(sources)
+      Mix.shell.info "mix test --stale did not find any tests to run, add --force if you want to run all tests"
```

Comment:
I am not sure. It would be nice to force the stale tests to run as a way to recompute the manifest. Something like `--stalef` could be nice because I would press "up" and type "f" but I am not sold on it. `--stale-force` could work too. Ideas?